languages/rust/exercises/concept/tuples/.docs/instructions.md

```diff
@@ -0,0 +1,48 @@
+# Instructions
+
+You are working on a game targeting a low-power embedded system and need to write several convenience functions which will be used by other parts of the game.
+
+## 1. Write a function `divmod` which returns both the quotient and remainder of a division
+
+```rust
+fn divmod(dividend: i16, divisor: i16) -> (i16, i16)
+```
+
+Example:
+
+```rust
+assert_eq!(divmod(10, 3), (3, 1));
+```
+
+## 2. Write an iterator adaptor `evens` which returns the even items from an arbitrary iterator
+
+This will be helpful to enable a screen-buffer optimization, your boss promises.
+
+[Iterators](https://doc.rust-lang.org/std/iter/trait.Iterator.html) are a separate concept, and we aren't going to get too in-depth about their syntax or mechanics here. The important facts about them for this task:
+
+- An iterator is an arbitrary-length stream of items
+- They have an [`enumerate` method](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate) which returns a tuple `(i, val)` for each value
+- They have a [`filter` method](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter) which uses a closure to determine whether to yield an element of the iterator
+- They have a [`map` method](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map) which uses a closure to modify elements of the iterator
+
+Don't worry too much about the generic or impl syntax here; it doesn't affect your implementation. It just means that this adaptor can be applied to any iterator, and the resulting iterator has the same item type as the input iterator.
+
+```rust
+fn evens<T>(iter: impl Iterator<Item=T>) -> impl Iterator<Item=T>
+```
```

Comment:
I think it's legit to say "don't worry too much about the signature; here are some things you can use to handle this", particularly given the hints, but I'll readily admit that this is probably the roughest part of the exercise. Suggestions welcome.

Reply:
I'll get to your question later, but this example jumped out at me
```rust
(0..).map(|n| 2 * n + 1).map(|n| n * n)
```
If the student plays with it, it could cause an issue
```rust
pub fn main() {
    let idgit = (0..).map(|n| 2 * n + 1).map(|n| n * n);
    let odbot = idgit.collect::<Vec<i32>>();
    println!("{:?}", odbot);
}

```
```
thread 'main' panicked at 'capacity overflow', library\alloc\src\raw_vec.rs:537:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
I think it might be better to limit it
```rust
(0..10).map(|n| 2 * n + 1).map(|n| n * n)
```

Reply:
Wow, there is a lot of thought and work you put into this.

As for a function which involves iterators along with tuples, I think something simpler like the existing Raindrops is a better situation in which to introduce tuples. For example,

```rust
pub fn raindrops(n: u32) -> String {
    let factors = [(3, "Pling"), (5, "Plang"), (7, "Plong")];
    let output = factors
        .iter()
        .map(|factor| if n % factor.0 == 0 { factor.1 } else { "" })
        .collect::<String>();

    match output.is_empty() {
        true => n.to_string(),
        false => output,
    }
}
```
Then again, the very first exercise `Hello World` tells the student to disregard the `'static` lifetime annotation, and that doesn't seem to pose a problem.

Reply:
Well, that depends on what you need. If you just need an _iterator_, then the form I specified is correct. If what you need is the _first N_, then you want either the form you suggested or `.take(n)`. The reason I chose an unbounded iterator is because that's the simplest to describe in human, mathematical language: an iterator of odd squares. 

I'm not sure it makes sense to worry too much about students abusing that example. If they haven't heard of iterators before, they can just use it, as was intended, as an example of closure literals. If they have heard of iterators before, then a moment's thought demonstrates why it's not a great idea to try to `.collect()` an unbounded iterator. In neither case have we introduced too much unnecessary confusion.

Reply:
The goal isn't necessarily to involve iterators with tuples; iterators are a distinct concept that should ideally be taught later. The goal is to introduce a simple situation where tuple destructuring is normal and expected; it just happens that AFAICT, iterators are the first case of that for most people.

Reply:
I think I see what you mean. Maybe a better existing example for introducing tuples might be the Bob exercise
```rust
pub fn reply(message: &str) -> &str {
    match (
        message.trim_end().is_empty(),
        message.trim_end().ends_with("?"),
        message == message.to_uppercase(),
        message.chars().any(|char| char.is_alphabetic()),
    ) {
        (true, _, _, _) => "Fine. Be that way!",
        (_, true, true, true) => "Calm down, I know what I'm doing!",
        (_, false, true, true) => "Whoa, chill out!",
        (_, true, _, _) => "Sure.",
        _ => "Whatever.",
    }
}
```

Reply:
Hm. That's an interesting idea, but I see two problems with it:

- That teaches destructuring matching, instead of destructuring binding; the latter is IMO the more common case. I want to teach that explicitly.
- There are a _lot_ of ways to complete `bob`, which disqualifies it from being a concept exercise. We want to design these such that an automated checker can say either "good, you found the ideal solution" or "try again" without leading to students giving up in frustration. This implies structuring the exercise such that the only reasonable solution is the optimal one.

Reply:
Right, that's a good point about the auto-checking, and also about the destructured binding. I'm reading the v3 intro docs now, as well as watching the video on concepts, so I hope my input may be more relevant as I get more grounded in the v3 mindset. Thanks!

Reply:
Hey folks!
@coriolinus  I love your idea to use a game development story with some maths. I think that works quite well to introduce tuples.
I do not feel confident working with iterator adapters yet and I wonder if it could overwhelm new students.
I see @bobahop's point in raising the raindrops and bob example.

For future iterations: Can y'all think of a way we might extract the best of raindrops or Bob for the second aspect of this task? I'm wondering how we can introduce tuples without iterators.

Reply:
>  I wonder if it could overwhelm new students

That's why I explicitly linked to the documentation for the `map`, `filter`, and `enumerate` methods. The students don't need more than that.

Reply:
> Maybe a better existing example for introducing tuples might be the Bob exercise

FYI it is not allowed to convert an existing practice exercise to a concept exercise. It is of course perfectly fine to take inspiration from existing exercises.

Reply:
> it is not allowed to convert an existing practice exercise to a concept exercise.

Why not? It's not relevant to this PR, but later on, there's an existing `macros` practice exercise which was designed from the start to teach the "macros-by-example" concept, before v3 was a thing. Why shouldn't that be convertable?

Reply:
The main reason behind this guideline is that virtually none of the existing exercises is designed to teach a single concept. The macros exercise sounds like it is the exception.